[MRG] Implement support for Individual Conditional Expectation (ICE) plots

madhuracj opened the pull request:
This PR adds support for Individual Conditional Expectation (ICE) plots

#### Reference Issues/PRs
Fixes #14126 

#### What does this implement/fix? Explain your changes.
##### API
Following how the partial dependence is implemented, this PR adds the following methods and classes to the public API
`sklearn.inspection.individual_conditional_expectation`,
`sklearn.inspection.plot_individual_conditional_expectation`,
`sklearn.inspection.IndividualConditionalExpectationDisplay`.

##### Implementation
Since partial dependence calculation is closely related to ICE, relevant common methods have been moved to and common code segments have been extracted into `_base.py`. 

##### Tests
Since PDP and ICE share quite a bit of code, most of the code in `_ice.py` is already covered by existing unit tests. Common unit tests have been moved to `test_plot_pdp_ice.py`. Additional unit testing has been added in `test_plot_ice.py` and `test_ice.py`. 

##### Examples and Doc
Examples on how to plot ICEs, centered ICEs and comparison of ICEs with PDPs have been added.

#### Any other comments?
Would appreciate reviews and suggestions.

madhuracj committed:
Implement individual_conditional_expectation, plot_individual_conditional_expectation methods and IndividualConditionalExpectationDisplay class

madhuracj committed:
Move _grid_from_X related tests to test_ice.py as they are now in _ice.py

madhuracj committed:
Add a test for the validation of 'features' parameter

madhuracj committed:
Tes shape of the predictions output as it has an additional dimension compared to partial_dependence

madhuracj committed:
Remove unused imports

madhuracj committed:
Fix lint errors on line length and indentation

madhuracj committed:
Fix line length

madhuracj committed:
Move tests applicable to plot_pdp and plot_ice to a new file and add coverage for plot_ice as well.

madhuracj committed:
Move common tests to test_plot_pdp_ice

madhuracj committed:
Move another test that is applicable to both plot_pdp and plot_ice

glemaitre commented on sklearn/inspection/_ice.py:
```diff
@@ -0,0 +1,832 @@
+           'IndividualConditionalExpectationDisplay']
+
+
+def _grid_from_X(X, percentiles, grid_resolution):
```
I cannot review right now but it seems really redundant. PDP and ICE should share code and we should factorize code as much as possible. you can create a `_base.py` file to import common part.

madhuracj replied:
Thanks, @glemaitre. I have extracted a number of common code into methods inside `_ice.py` which are being used by both PDP (https://github.com/madhuracj/scikit-learn/blob/06b1c617f9a28d59822d4346d1087d652ef9631d/sklearn/inspection/_partial_dependence.py#L16-L18) and ICE. 

For example, see https://github.com/madhuracj/scikit-learn/blob/06b1c617f9a28d59822d4346d1087d652ef9631d/sklearn/inspection/_partial_dependence.py#L188-L191 and https://github.com/madhuracj/scikit-learn/blob/06b1c617f9a28d59822d4346d1087d652ef9631d/sklearn/inspection/_ice.py#L353-L356.

I agree it makes more sense to have them in `_base.py` rather than `_ice.py`. Will update the PR with that.

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/2da2d4ccc1e668d44f589a4e45eae1440e91b474

madhuracj committed:
Move common methods to _base.py

madhuracj committed:
Use a more representative name for the parameter

madhuracj committed:
Allow centering ICE curves

madhuracj committed:
Add examples for ICE plots

madhuracj committed:
Remove additinal blank lines at the end of the file

madhuracj committed:
Fix a couple of long lines

madhuracj commented:
A gentle reminder.
Appreciate if a maintainer could skim through and let me know if there is anything missing so I can work on that.

jnothman commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -5,159 +5,33 @@
 def _partial_dependence_brute(est, grid, features, X, response_method):
     averaged_predictions = []
-
-    # define the prediction_method (predict, predict_proba, decision_function).
```
To the reviewer: this is now _base._get_predictions

madhuracj replied:
This is now _partial_dependence._get_predictions as ICE was moved to _partial_dependence.py file.

jnothman commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -5,159 +5,33 @@
            'PartialDependenceDisplay']
 
 
-def _grid_from_X(X, percentiles, grid_resolution):
```
moved to _base

madhuracj replied:
Moved back to _partial_dependence.py as suggested.

jnothman commented on sklearn/inspection/_base.py:
```diff
@@ -0,0 +1,365 @@
+    return predictions
+
+
+def _validate_pdp_ice_parameters(estimator, X, features, response_method,
```
Given how similar PDP and ICE are, and how different they are from other (prospectively) things in `inspection`. I would be tempted to implement ICE in the same file as PDP, allowing us to avoid this "pdp_ice" nomenclature.

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/154004e41588c4fb9615f408dd4a9cee24b280ca

jnothman commented on sklearn/inspection/_base.py:
```diff
@@ -0,0 +1,365 @@
+                  n_cols=3, grid_resolution=100, percentiles=(0.05, 0.95),
+                  method='auto', n_jobs=None, verbose=0, fig=None,
+                  line_kw=None, contour_kw=None, ax=None, **kwargs):
+
```
it might be helpful to have a docstring that describes what the role of this helper is, and particularly they key params that differentiate between ICE and PDP

madhuracj replied:
Thanks for the suggestion. Will add that.

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/feae02c7d2cc94ea10a2fd96b8a2a145bf14c891

jnothman commented on sklearn/inspection/_ice.py:
```diff
@@ -0,0 +1,491 @@
+"""Individual Conditional Expectation (ICE) plots for regression and
+classification models."""
+
+# Authors: Peter Prettenhofer
```
do you really think this file was such a broad collaboration?

madhuracj replied:
Before @glemaitre suggested using a _base.py I had moved/extracted common functions into _ice.py. Just wanted to give credit to the original authors of those code segments. I should have moved the credit to _base.py.

jnothman commented on sklearn/inspection/_ice.py:
```diff
@@ -0,0 +1,491 @@
+    ICE of a feature (or a set of features) corresponds to the responses of an
+    estimator for each possible value of the feature for all instance in ``X``.
+
+    Read more in the :ref:`User Guide <individual_conditional_expectation>`.
```
This user guide is not yet written, and will require you to update `doc/modules/partial_dependence.rst` perhaps renaming it

madhuracj replied:
Updated in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/b04ba687c014b1ba9beaa63724574cc330d975c9. I feel 'Partial dependence' covers both PDPs and ICEs with PDP showing the average dependence while ICE visualizing the dependence for each instance separately.

jnothman commented on sklearn/inspection/_ice.py:
```diff
@@ -0,0 +1,491 @@
+        The number of equally spaced points on the axes of the plots, for each
+        target feature.
+
+    centre : bool, optional (default=True)
```
this parameter is in a different order relative to percentiles in `individual_conditional_expectation`

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/a4e676208745edfd7058524dcab8d49d72598674

jnothman commented on examples/inspection/plot_individual_conditional_expectation.py:
```diff
@@ -0,0 +1,159 @@
+fig.subplots_adjust(hspace=0.3)
+
+
+# In ICE plots it might not be easy to see the average effect of the 'target'
```
This comment appears before the figure in the output (https://89544-843222-gh.circle-artifacts.com/0/doc/auto_examples/inspection/plot_individual_conditional_expectation.html#sphx-glr-auto-examples-inspection-plot-individual-conditional-expectation-py) and looks a bit out of place there.

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/0d9dafab8bd28297bb313e0eb701f101ea560786

jnothman commented on sklearn/inspection/_ice.py:
```diff
@@ -0,0 +1,491 @@
+        The number of equally spaced points on the grid, for each target
+        feature.
+
+    centre : bool, optional (default=True)
```
```suggestion
    center : bool, optional (default=True)
```

jnothman replied:
Maybe this can have a more explicit name, like "start_at_origin" or "fixed_start_point"

madhuracj replied:
I went ahead with `fixed_start_point`. Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/fee74e12fb4d6519c4f0fa42c975553d845bd5bc

jnothman commented on sklearn/inspection/_ice.py:
```diff
@@ -0,0 +1,491 @@
+        feature.
+
+    centre : bool, optional (default=True)
+        Whether to centre the ICE curves at the beginning of x-axes.
```
I think this could be described more clearly, but we can review the wording after there's a user guide.

madhuracj replied:
I hope I have described this bit better in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/fee74e12fb4d6519c4f0fa42c975553d845bd5bc

jnothman commented on sklearn/inspection/_ice.py:
```diff
@@ -0,0 +1,491 @@
+
+    X : {array-like or dataframe} of shape (n_samples, n_features)
+        ``X`` is used both to generate a grid of values for the
+        ``features``, and to compute the averaged predictions.
```
Is "averaged predictions" relevant to ICE?

madhuracj replied:
Fixed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/cb37159703cc553996ab87e269128408e8ee31b1

jnothman commented on sklearn/inspection/_ice.py:
```diff
@@ -0,0 +1,491 @@
+          >>> X, y = make_friedman1()
+          >>> est = LinearRegression().fit(X, y)
+          >>> disp1 = plot_individual_conditional_expectation(est,
+          ...                                 X)  # doctest: +SKIP
```
indentation here is awkward. Rather, put a line break after (

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/a3d6a634245918bf28f75758fdadc6df893ea197

jnothman commented on sklearn/inspection/tests/test_ice.py:
```diff
@@ -0,0 +1,132 @@
+     (100, (.9, .1), r"percentiles\[0\] must be strictly less than"),
+     (1, (0.05, 0.95), "'grid_resolution' must be strictly greater than 1")]
+)
+def test_grid_from_X_error(grid_resolution, percentiles, err_msg):
```
What is the benefit of duplicating this test of a private function?

madhuracj replied:
Sorry, this was an oversight when moving testing code around. Fixed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/154004e41588c4fb9615f408dd4a9cee24b280ca

jnothman commented on sklearn/inspection/tests/test_ice.py:
```diff
@@ -0,0 +1,132 @@
+                                               random_state=0), 2)
+
+
+def test_grid_from_X():
```
What is the benefit of duplicating this test of a private function?

madhuracj replied:
Sorry, this was an oversight when moving testing code around. Fixed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/154004e41588c4fb9615f408dd4a9cee24b280ca

jnothman commented on sklearn/inspection/tests/test_ice.py:
```diff
@@ -0,0 +1,132 @@
+@pytest.mark.parametrize('grid_resolution', (5, 10))
+@pytest.mark.parametrize('features', ([1], [1, 2]))
+def test_output_shape(Estimator, data, grid_resolution, features):
+    # Check that individual_conditional_expectation has consistent output
```
Could we add more tests of ICE output on toy data, not merely the shape?

madhuracj replied:
Sorry, this was an oversight when moving testing code around. Fixed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/154004e41588c4fb9615f408dd4a9cee24b280ca

madhuracj replied:
The actual ICE output is tested in 
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/7f673f9cbb0a2e3a60847d6a64fcd8a6119d0a78

jnothman reviewed (commented):
Hello from Sydney, @madhuracj!

Thanks for the PR and for the ping. I will admit that one big pity about doing this in matplotlib is that I would love to be able to mouse over the curve for a sample in one plot, and see the corresponding sample highlighted on the other subplots. Oh well.

This is only a partial review. I've not looked closely at the plot method or some tests.

I don't mind the documentation more explicitly presenting both ICE and Partial Dependence under the same heading.

In general, I'd be happier with fewer files, esp for testing...

I'm wondering about the nomenclature, "lines". Should we be calling these "curves"?

madhuracj commented:
Hello @jnothman,

Appreciate you taking time on your holiday to review the PR. 

>I don't mind the documentation more explicitly presenting both ICE and Partial Dependence under the same heading.
>In general, I'd be happier with fewer files, esp for testing...

>Given how similar PDP and ICE are, and how different they are from other (prospectively) things in inspection. I would be tempted to implement ICE in the same file as PDP, allowing us to avoid this "pdp_ice" nomenclature.

I like the idea of implementing ICE in the same file as PDP. However, I cannot seem to think of a name to represent both PDP and ICE, so I'll leave the file as `_partial_dependence.py` for now. I agree this will be much cleaner without multiple files for unit tests, examples, documentation etc.

madhuracj committed:
Reduce the number of files

Implement ICE inside _partial_dependence.py. Move ICE tests to test_partial_dependence.py and test_plot_partial_dependence.py.

madhuracj committed:
Add a docstring describing the role of the helper function.

madhuracj committed:
Use the same parameter order as individual_conditional_expectation

madhuracj committed:
Move the comment down so it appears after the figure

madhuracj committed:
Use a more explicit name for the parameter. Better description for the parameter.

madhuracj committed:
Averages predictions are only for PDPs

madhuracj committed:
Fix indentation

madhuracj committed:
Merge branch 'master' into ice

madhuracj committed:
plot_ice and ice methods will be added in 0.23

madhuracj committed:
Use the updated parameter name 'fixed_start_point'

glemaitre commented on sklearn/inspection/__init__.py:
```diff
@@ -12,14 +12,20 @@
     warnings.simplefilter("ignore", category=FutureWarning)
     from .partial_dependence import partial_dependence
 
+from ._partial_dependence import individual_conditional_expectation  # noqa
```
I think that we don't need to have `#noqa` since they are added in `__all__`

madhuracj replied:
Thanks, I have removed the #noqa annotations.

madhuracj replied:
Removing these annotations attract the linting error 'E402 module level import not at top of file'. Reverting...

madhuracj committed:
Use 'curves' inplace of 'lines'

madhuracj committed:
Fix linting issues

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -137,27 +140,33 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+def _partial_dependence_brute(est, grid, features, X, response_method):
+    averaged_predictions = []
+    predictions = _get_predictions(est, grid, features, X, response_method)
+    for prediction in predictions:
```
```suggestion
    for pred in predictions:
```

madhuracj replied:
Added in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/384d6b7f6822b431894aa776a4892f7829eb50dc

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -137,27 +140,33 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+    for prediction in predictions:
         # average over samples
-        averaged_predictions.append(np.mean(predictions, axis=0))
+        averaged_predictions.append(np.mean(prediction, axis=0))
```
```suggestion
        averaged_predictions.append(np.mean(pred, axis=0))
```

madhuracj replied:
Added in Added in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/384d6b7f6822b431894aa776a4892f7829eb50dc

madhuracj committed:
Use variable 'pred' as suggested

madhuracj committed:
'noqa' is not required as imports are being used

glemaitre commented:
@mabdelaal86 I will review soon but I wanted to check something.

From what I am recalling, PD is an average of the ICE. I was wondering if we can even factorize more the code by calling ICE in PD?

madhuracj committed:
Revert 7933a93

madhuracj committed:
Unit tests for the actual output of individual_conditional_expectation method

madhuracj committed:
Improve wording, formatting of the ICE example

madhuracj committed:
Use the same layout as PDP for easy comparison

madhuracj committed:
Update partial dependence user guide with ICE details

madhuracj commented:
@jnothman I have completed the suggested changes. Appreciate if you could have a look sometime.

madhuracj committed:
Fix math formatting

madhuracj committed:
Improve documentation

madhuracj committed:
Add new methods and class to the API reference

madhuracj committed:
Fix indentation

madhuracj committed:
Add missing docs for arguments

thomasjpfan commented:
I would prefer this to be in PD with a keyword parameter. Maybe `average=None, 'mean'`, where the default is `'mean'`.

madhuracj committed:
Fix links to the user guide

madhuracj commented:
Thanks, @thomasjpfan. I considered a similar approach initially, however, revered to the current API specification as ICEs seems to be identified specifically by their name and separately from PDPs. For example, in [1] two sections are devoted to the two types of plots. Implementation wise, they share quite a lot of common code internally. For example, see [2] and [3]

[1] https://christophm.github.io/interpretable-ml-book
[2] https://github.com/scikit-learn/scikit-learn/blob/0fa426c083af1e7adeee72dc1fb134d756ce445d/sklearn/inspection/_partial_dependence.py#L868-L874
[3] https://github.com/scikit-learn/scikit-learn/blob/0fa426c083af1e7adeee72dc1fb134d756ce445d/sklearn/inspection/_partial_dependence.py#L1382-L1388

madhuracj commented:
@jnothman @glemaitre Do you also agree that using PD with a keyword parameter is best in terms of the API specification?

jnothman commented:
Give me a usage example, please?

madhuracj commented:
@jnothman I am not sure whether I understood your question correctly. 
Do you mean an instance where PDPs and ICEs are exposed as different methods in some other library? See http://austinrochford.github.io/PyCEbox/docs/#pycebox.ice.ice

jnothman commented:
No I mean, what does it look like to the user or code maintainer under each
proposal?

glemaitre commented:
@thomasjpfan I don't agree with your API choice.

While it makes sense from a programmer perspective, it will be less intuitive for users to discover this feature. I assume that people will google ICE or search ICE in our documentation and having it hidden behind a parameter named `average` in a class/function named `partial_dependence` will be challenging.

I am more inclined to refactoring the code such that we call ICE in PDP but having 2 classes/functions exposed to the end-user to be aligned with the naming convention in the literature (as mentioned by @madhuracj https://github.com/scikit-learn/scikit-learn/pull/16164#issuecomment-580463450)

IRL @ogrisel is also sharing this idea of separated functions/classes.

We should probably agree on this higher-level API before to make further reviews.

@thomasjpfan Could you provide some argument regarding your API choice such that we can resolve go forward?

thomasjpfan commented:
I am happy with having a new method/class for discoverability.

madhuracj commented:
@jnothman Find below how it would look like to the user, 
Proposed approach in this PR:
PDP
```
partial_dependence(estimator, X, features, ...)
plot_partial_dependence(estimator, X, features, ...)
```
ICE
```
individual_conditional_expectation(estimator, X, features, ...)
plot_individual_conditional_expectation(estimator, X, features, ...)
```

Alternative approach proposed by @thomasjpfan :
PDP
```
partial_dependence(estimator, X, features, average='mean', ...)
plot_partial_dependence(estimator, X, features, average='mean', ...)
```
ICE
```
partial_dependence(estimator, X, features, average='None', ...)
plot_partial_dependence(estimator, X, features, average='None', ...)
```

thomasjpfan commented:
Let's go with the `plot_individual_conditional_expectation` as it is in the PR.

jnothman commented:
Yeah I'd prefer "individual=True" instead of average=None, but two seperate
functions is okay by me.

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -25,14 +26,16 @@
 
-
-__all__ = ['partial_dependence', 'plot_partial_dependence',
+__all__ = ['individual_conditional_expectation',
```
We have 1500 lines of code now. I think this the moment to split things.

I would create a submodule called `_plot` and add the plotting facilities there.
We should also rename the file from `_partial_dependence` to something which is both working for PD and ICE.

Basically, we need a nice name for explaining that we deal with global inspection which check the marginal effect of a feature(s) with the target. So I don't know if we should call it `_global.py`. Everything that we have for the moment is indeed a global inspection. Maybe `_feature_interaction.py` might be better.

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        The number of equally spaced points on the grid, for each target
+        feature.
+
+    fixed_start_point : bool, optional (default=True)
```
I think this parameter is YAGNI. We will not need it at first. I prefer to be close to the implementation proposed by the original paper and have feedback from users before to introduce thing which we are not sure to be useful.

madhuracj replied:
Removed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/383ec6ceba008e06630d99c61820dfe721c58ba0

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+    return predictions
+
+
+def individual_conditional_expectation(estimator, X, features,
```
I think that we could add a useful parameter `subsample`. Basically, it will not be easy to visualize all responses and we need to subsample anyway. So we can add the parameter directly.

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+    fixed_start_point : bool, optional (default=True)
+        Whether to use a fixes starting point for all the ICE curves.
+
+    kwargs : dict
```
This should go away

madhuracj replied:
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/cb54fbd95fefb5a89a6b395c26f11b76ec4f1f9d

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+    (array([[[2.19017313e-05, 5.16319395e-01],
+        [4.83680605e-01, 9.99978098e-01]]]), [array([0., 1.])])
+    """
+    X, response_method, _ = _validate_parameters(estimator, X, features,
```
What is the reason for supporting only the `'brute'` method?

madhuracj replied:
https://github.com/scikit-learn/scikit-learn/pull/16164#discussion_r381018502

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        The target features for which to create the ICEs.
+        if any entry is a string, then it must be in ``feature_names``.
+
+    response_method : 'auto', 'predict_proba' or 'decision_function', \
```
```suggestion
    response_method : {'auto', 'predict_proba', 'decision_function'}, \
```

madhuracj replied:
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/ba8b58bfddaa65c749eeadbc5d47edf0e01c0ac7

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        if any entry is a string, then it must be in ``feature_names``.
+
+    response_method : 'auto', 'predict_proba' or 'decision_function', \
+            optional (default='auto')
```
```suggestion
            default='auto'
```

madhuracj replied:
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/ba8b58bfddaa65c749eeadbc5d47edf0e01c0ac7

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        :term:`predict`. By default, :term:`predict_proba` is tried first
+        and we revert to :term:`decision_function` if it doesn't exist.
+
+    percentiles : tuple of float, optional (default=(0.05, 0.95))
```
```suggestion
    percentiles : tuple of floats, default=(0.05, 0.95)
```

madhuracj replied:
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/ba8b58bfddaa65c749eeadbc5d47edf0e01c0ac7

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        The lower and upper percentile used to create the extreme values
+        for the ICE axes. Must be in [0, 1].
+
+    grid_resolution : int, optional (default=100)
```
```suggestion
    grid_resolution : int, default=100
```

madhuracj replied:
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/ba8b58bfddaa65c749eeadbc5d47edf0e01c0ac7

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        The number of equally spaced points on the axes of the plots, for each
+        target feature.
+
+    fixed_start_point : bool, optional (default=True)
```
to be removed

madhuracj replied:
Removed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/383ec6ceba008e06630d99c61820dfe721c58ba0

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        By default, the name of the feature corresponds to their numerical
+        index for NumPy array and their column name for pandas dataframe.
+
+    target : int, optional (default=None)
```
```suggestion
    target : int, default=None
```

madhuracj replied:
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/ba8b58bfddaa65c749eeadbc5d47edf0e01c0ac7

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+
+        Ignored in binary classification or classical regression settings.
+
+    n_cols : int, optional (default=3)
```
Basically remove all the optional occurrences in your added docstring

madhuracj replied:
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/ba8b58bfddaa65c749eeadbc5d47edf0e01c0ac7

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -467,23 +450,27 @@ def plot_partial_dependence(estimator, X, features, feature_names=None,
+    calc_method : partial_dependence or individual_conditional_expectation
+        The function to calculate values for the PDP/ICE plot
+
+    display_class : PartialDependenceDisplay or
```
Basically we don't need both `calc_method` and `display_class` since we don't want to make a mix `ice/pd`
Maybe we can introduce the proposed keyword by @jnothman `individual` which will be a bool.
Internally we will choose the method to call with the display associated.

madhuracj replied:
Only saw this now. I've used `plot_type` which is a bit more explicit I suppose. Happy to use `individual` if you find it better.

https://github.com/scikit-learn/scikit-learn/pull/16164/commits/cb54fbd95fefb5a89a6b395c26f11b76ec4f1f9d#diff-0c8623ee957e0256d29ac53830094281R453-R454

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -532,64 +519,48 @@ def plot_partial_dependence(estimator, X, features, feature_names=None,
 
-    verbose : int, optional (default=0)
-        Verbose output during PD computations.
+        verbose : int, optional (default=0)
```
no need to indent

madhuracj replied:
Fixed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/4a31914f72a50286e2248465c0641e4e38d5b337

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -532,64 +519,48 @@ def plot_partial_dependence(estimator, X, features, feature_names=None,
+        Additional keyword arguments passed to the ``calc_method`` and the
+        ``plot`` method of ``display_class``.
+        For ICEs, parameter ``fixed_start_point`` will be included in kwargs.
+        For PDPs, parameter ``contour_kw`` will be included in kwargs.
```
Basically, it will remain only `contour_kw` which can be directly given with a default value. It will only have an effect for the PDP plotting but this is fine.

madhuracj replied:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/cb54fbd95fefb5a89a6b395c26f11b76ec4f1f9d

glemaitre reviewed (commented):
The first round of comment in case GitHub shows me a unicorn if I am too much talkative.

glemaitre commented on sklearn/inspection/tests/test_plot_partial_dependence.py:
```diff
@@ -399,20 +407,55 @@ def test_plot_partial_dependence_dataframe(pyplot, clf_boston, boston):
+
+    with pytest.raises(ValueError, match=err_msg):
+        plot_partial_dependence(estimator, X, **params)
+    with pytest.raises(ValueError, match=err_msg):
```
a general remark for the test. If we expect the same behaviour for some part, let's just parametrize the test with the function to try. It will be more elegant with less repetitive code.

madhuracj replied:
Wouldn't this mean duplicating the whole parameter set (https://github.com/scikit-learn/scikit-learn/pull/16164/files/0fa426c083af1e7adeee72dc1fb134d756ce445d#diff-1206a47695776d1666e6e6873502b8e5R398-R415)?

glemaitre commented on sklearn/inspection/tests/test_plot_partial_dependence.py:
```diff
@@ -380,6 +384,10 @@ def test_plot_partial_dependence_dataframe(pyplot, clf_boston, boston):
         feature_names=df.columns.tolist()
     )
+    plot_individual_conditional_expectation(
+        clf_boston, df, ['TAX', 'AGE'], grid_resolution=grid_resolution,
```
We should switch from Boston to another dataset from now on. Since we are touching the test, feel free to take another one. I think that you can take another regression toy dataset for the purpose of these unit tests. We are only checking for operational behaviour and not results.

glemaitre commented on sklearn/inspection/tests/test_partial_dependence.py:
```diff
@@ -268,6 +312,19 @@ def test_partial_dependence_easy_target(est, power):
 
     assert r2 > .99
 
+    predictions, values = individual_conditional_expectation(
```
Either parametrize the test or just move the test in a new file dedicated to ICE.

glemaitre commented on examples/inspection/plot_individual_conditional_expectation.py:
```diff
@@ -0,0 +1,162 @@
+"""
+==============================================
```
I think that I would not introduce a new example but instead modify the existing PDP example. We could add a section on ICE to explain what is the intention when plotting the individual response and add a little paragraph when discussing the plot to show the usefulness of ICE in regards to the PDPs.

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+    >>> individual_conditional_expectation(gb, features=[0], X=X,
+    ...       percentiles=(0, 1), grid_resolution=2) # doctest: +SKIP
+    (array([[[2.19017313e-05, 5.16319395e-01],
+        [4.83680605e-01, 9.99978098e-01]]]), [array([0., 1.])])
```
We need to add the reference to the original paper (which you earlier added in your example).

madhuracj replied:
Added in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/62d112b230c3d66e922bf5c95d4c80ef126e4640

glemaitre commented on doc/modules/partial_dependence.rst:
```diff
@@ -1,21 +1,25 @@
-Partial dependence plots
-========================
+==================
+Partial dependence
```
I think that we need  to change the name of the section

Maybe "Interaction between feature and target"

and then add 2 section PDP and ICE

glemaitre commented on doc/modules/partial_dependence.rst:
```diff
@@ -1,21 +1,25 @@
 .. currentmodule:: sklearn.inspection
 
-Partial dependence plots (PDP) show the dependence between the target
+Partial dependence plots (PDP) and individual conditional
```
I really think that we should separate both approaches. Maybe, it will make sense to add first ICE and then PDP and explain that this the average response. I fell that we might miss a small "mathematical formulation" section. It could be a good addition

glemaitre reviewed (commented):
I feel this is enough comments to make quite a lot of changes. I would make another review once files are split and the example are slightly change, we can forward then. But it looks promising.

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -177,9 +186,98 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+        # assuming Pipeline is the only estimator that does not store a new
+        # attribute
+        for est in estimator:
+            # FIXME: remove the None option when it will be deprecated
```
Added in https://github.com/scikit-learn/scikit-learn/pull/14028/files#diff-0c8623ee957e0256d29ac53830094281R315.

The deprecation seems to be in `FeatureUnion` rather than `Pipeline`. See https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/pipeline.py#L773-L774

madhuracj committed:
Do not use kwargs

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -262,6 +360,9 @@ def partial_dependence(estimator, X, features, response_method='auto',
         Please see :ref:`this note <pdp_method_differences>` for
         differences between the 'brute' and 'recursion' method.
 
+    kwargs : dict
```
Fixed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/cb54fbd95fefb5a89a6b395c26f11b76ec4f1f9d

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -467,23 +450,27 @@ def plot_partial_dependence(estimator, X, features, feature_names=None,
+        The target features for which to create the PDP/ICEs.
         if any entry is a string, then it must be in ``feature_names``.
 
+    calc_method : partial_dependence or individual_conditional_expectation
```
No longer used after https://github.com/scikit-learn/scikit-learn/pull/16164/commits/cb54fbd95fefb5a89a6b395c26f11b76ec4f1f9d

madhuracj committed:
Remove 'optional' in parameter documentation.

madhuracj committed:
Linting. Fix line length

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        The feature (e.g. `[0]`) or pair of interacting features
+        (e.g. `[(0, 1)]`) for which the ICE should be computed.
+
+    response_method : 'auto', 'predict_proba' or 'decision_function', \
```
Fixed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/ba8b58bfddaa65c749eeadbc5d47edf0e01c0ac7

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        :term:`predict`. By default, :term:`predict_proba` is tried first
+        and we revert to :term:`decision_function` if it doesn't exist.
+
+    percentiles : tuple of float, optional (default=(0.05, 0.95))
```
Fixed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/ba8b58bfddaa65c749eeadbc5d47edf0e01c0ac7

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+        The lower and upper percentile used to create the extreme values
+        for the grid. Must be in [0, 1].
+
+    grid_resolution : int, optional (default=100)
```
Fixed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/ba8b58bfddaa65c749eeadbc5d47edf0e01c0ac7

madhuracj committed:
Remove 'fixed_start_point' parameter

madhuracj committed:
Fix indentation

madhuracj committed:
Fix output of the ice function

glemaitre commented:
Regarding the failure, you only need to merge master into your branch. It should not be any issue since you don't have any conflict.

glemaitre commented:
I opened a PR to split the partial dependence computation and the visualization: https://github.com/scikit-learn/scikit-learn/pull/16453

My hope is to merge and change the name of the submodule such that it will be much easier to integrate your changes and review them.

madhuracj committed:
Merge branch 'master' into ice

madhuracj committed:
Add a reference to the original paper

madhuracj commented:
I will wait for #16453 to be merged before additional changes.

jnothman commented:
> I will wait for #16453
<https://github.com/scikit-learn/scikit-learn/pull/16453> to be merged
before additional changes.

Proceed!

glemaitre commented:
I will solve the conflict and push in your branch. We did a bit of mess there :)

glemaitre committed:
Merge remote-tracking branch 'origin/master' into pr/madhuracj/16164

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -25,10 +31,12 @@
 from sklearn.ensemble._hist_gradient_boosting.gradient_boosting import (
     BaseHistGradientBoosting)
 
-
```
We need to add a docstring

```
def_get_predictions(...):
    """Get predictions for an estimator by replacing the values of
    a specific feature, keeping others constant.

    Parameters
    ----------
    estimator : estimator object
        The scikit-learn predictor.
    grid : ndarray of shape (n_points, n_target_features)
        The different values to try for the targeted features.
    features_indices : ndarray of shape (n_targeted_features,)
        The indices of the features targeted. The values of these
        features will be changed.
    X : {ndarray, dataframe} of shape (n_samples, n_features)
        The dataset to be used.
    response_method : {"auto", "predict_proba", "decision_function"}
        The response method to use by the estimator. It applies only for classifier.
    """
```

We should rename est to estimator

glemaitre replied:
The shape of the array need to be checked to be sure.

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/58630130d1639431591c7b75773466099ab75983.

Parameter validation is already happening at https://github.com/scikit-learn/scikit-learn/blob/58630130d1639431591c7b75773466099ab75983/sklearn/inspection/_partial_dependence.py#L229-L230
and 
https://github.com/scikit-learn/scikit-learn/blob/58630130d1639431591c7b75773466099ab75983/sklearn/inspection/_partial_dependence.py#L281-L288
Do you want any additional validation to be performed?

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -132,27 +133,33 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+            predictions.append(prediction_method(X_eval))
         except NotFittedError:
             raise ValueError(
                 "'estimator' parameter must be a fitted estimator")
```
```suggestion
                "'estimator' parameter must be a fitted estimator"
            )
```

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -172,6 +179,95 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
     return averaged_predictions
 
 
+def _validate_parameters(estimator, X, features, response_method, method):
```
Let's move the validation on the top of the file. Then _grid_from_X and the the different _partial_dependence and _ice_brute

madhuracj replied:
Moved in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/640f938124fd8442fdee4c3e32b0cb2cc108fbbc

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -172,6 +179,95 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+    accepted_responses = ('auto', 'predict_proba', 'decision_function')
+    if response_method not in accepted_responses:
+        raise ValueError(
+            'response_method {} is invalid. Accepted response_method names '
```
```suggestion
           f 'response_method {response_method} is invalid. Accepted "response_method" names '
```

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/b9d4a502b0cc31f05c22cada4f1a17bb91e35dd3

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -172,6 +179,95 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+    if response_method not in accepted_responses:
+        raise ValueError(
+            'response_method {} is invalid. Accepted response_method names '
+            'are {}.'.format(response_method, ', '.join(accepted_responses)))
```
```suggestion
            f'are {','.join(accepted_responses)}.' ))
```

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/b9d4a502b0cc31f05c22cada4f1a17bb91e35dd3

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -172,6 +179,95 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+    accepted_methods = ('brute', 'recursion', 'auto')
+    if method not in accepted_methods:
+        raise ValueError(
+            'method {} is invalid. Accepted method names are {}.'.format(
```
```suggestion
            f'The method {method} is invalid. Accepted "method" names are {', '.join(accepted_methods)} method.'
```

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/b9d4a502b0cc31f05c22cada4f1a17bb91e35dd3

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -172,6 +179,95 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+        else:
+            method = 'brute'
+
+    if method == 'recursion':
```
Uhm it should be an `elif` here otherwise we will test even after setting for the `"auto"` case

madhuracj replied:
'method' is resolved from 'auto' to 'brute' or 'recursion' in https://github.com/scikit-learn/scikit-learn/blob/b9d4a502b0cc31f05c22cada4f1a17bb91e35dd3/sklearn/inspection/_partial_dependence.py#L84-L90.

Subsequent validations for `method = 'recursion'` needs to be performed I guess.

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -172,6 +179,95 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+                'HistGradientBoostingRegressor',
+            )
+            raise ValueError(
+                "Only the following estimators support the 'recursion' "
```
Use an f-string

madhuracj replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/2894ebe79543504a5cbfe06382dc3ff6effa17a8

glemaitre commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -172,6 +179,95 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+                "Only the following estimators support the 'recursion' "
+                "method: {}. Try using method='brute'."
+                .format(', '.join(supported_classes_recursion)))
+        if response_method == 'auto':
```
Jump a line

madhuracj replied:
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/5bb429ef0d996415ebf3c466f7cbe3259f39bfdf

glemaitre commented on sklearn/inspection/_plot/partial_dependence.py:
```diff
@@ -274,14 +221,23 @@ def convert_feature(fx):
-                                    percentiles=percentiles)
-        for fxs in features)
+    # compute predictions
+    if plot_type == 'pdp':
```
We need the name of the common parameters to be the same and then we can simplify like:

```python
params = {
    "est": estimator, "X": X, ...
}
if plot_type == partial_dependence:
    params["method"] = method
    pd_results = Parallel(n_jobs=n_jobs, verbose=verbose)(
            delayed(plot_type)(**params) for fxs in features
    )
```

glemaitre commented on sklearn/inspection/_plot/partial_dependence.py:
```diff
@@ -322,10 +278,205 @@ def convert_feature(fx):
-                                       target_idx, pdp_lim, deciles)
-    return display.plot(ax=ax, n_cols=n_cols, line_kw=line_kw,
-                        contour_kw=contour_kw)
+    if plot_type == 'pdp':
```
Same trick here than for the `Parallel` using a `dict`

glemaitre reviewed (commented):
That's it for tonight. I will continue tomorrow :)

NicolasHug commented on doc/modules/partial_dependence.rst:
```diff
@@ -1,21 +1,25 @@
-Partial dependence plots (PDP) show the dependence between the target
+Partial dependence plots (PDP) and individual conditional
+expectation (ICE) plots show the dependence between the target
 response [1]_ and a set of 'target' features, marginalizing over the values
```
The ICE plots don't marginalize so we should re-formulate that sentence as well

NicolasHug commented on doc/modules/partial_dependence.rst:
```diff
@@ -57,38 +81,68 @@ and a two-way PDP between the two features::
+            [0.99704493, 0.99704493, 0.99704493, ..., 0.99051952,
+             0.99806429, 0.99806429]]])
+    >>> axes
+    [array([-1.624..., -1.592..., ...
```
I think it's OK not to write the example here, considering that you properly mention the existence of the function before, and its usage is similar to `partial_dependence` which is illustrated. (Also, I don't expect these functions to be really used by anyone)

madhuracj replied:
Removed in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/61467e80861895176a8fc0533c4e79de241bbbea

NicolasHug commented on doc/modules/partial_dependence.rst:
```diff
@@ -131,7 +189,9 @@ approximates the above integral by computing an average over the data `X`:
-over the dataset `X` which is computationally intensive.
+over the dataset `X` which is computationally intensive. Note that only
+'brute' method is available for individual conditional expectation
+calculation.
```
```suggestion
calculation, since the averaging is already implicit in the 'recursion' method.
```

NicolasHug reviewed (commented):
Thanks for the PR @madhuracj , this will be a nice addition.

I haven't given this a thorough look but I'll try to review soonish. 

In terms of API, I think it would be more natural to just add a parameter to `plot_partial_dependence()`: `plot_partial_dependence(average=True)` would be the default (PDP) and `average=False` would be the ICE plots. We could also have `average='with_ICE'` which would plot both the PDP *and* the ICE, as in the original paper, which would be useful in general and for illustration in the examples.

glemaitre commented:
@NicolasHug We discuss this matter a bit earlier in the discussion: https://github.com/scikit-learn/scikit-learn/pull/16164#issuecomment-585755407

In short, I think this is best to have separated functions/classes for a user to discover it in the documentation while searching since the literature refer ICE with a specific name. I also agree that having a parameter would make things neat, API and code-wise, but I would favour the current proposal.

NicolasHug commented:
Thanks, I couldn't find the past discussion.

I agree discoverability is an issue, but I think we can address that better by having carefully designed docs rather than adding a new function? E.g. change the title to "Partial Dependence and ICE plots", have a whole example dedicated to ICE, make it clear from the docstring, etc.

Having a single function isn't just practical from a design POV but it also highlights the fact that ICEs are merely the PD bits that we don't average, which is valuable from a pedagogical perspective.

Anyway, no strong opinion.

I do think however we should support both PDPs *and* ICEs on the same plot, and I'm not sure how that could look like with 2 separate functions

glemaitre commented:
> I do think however we should support both PDPs and ICEs on the same plot, and I'm not sure how that could look like with 2 separate functions

Yes, I agree. I think that we should be able to pass the matplotlib axis around.

NicolasHug commented:
How do we avoid re-computation though? This is important because we can only rely on the brute method which can be pretty slow

glemaitre commented:
> How do we avoid re-computation though?

Yep, this is the bottleneck. Maybe the only way would be to pass the ICE output to the PDP function (and I am not thinking about the plotting functions yet) instead of the estimator and data. Then, I don't know if we start to have an issue regarding the cumbersome API. Do you have any other idea to bypass the issue?

I am thinking to write a quick SLEP-like thing with the 2 possibilities, their pros and cons and to get input from other devs. I am not sure that we have a clear best choices.

NicolasHug commented:
> Do you have any other idea to bypass the issue?

Apart from having 1 function, no ^^

+1 for a RFC (SLEP might be overkill?). We can also plan a discussion for the meeting Monday


(Just to clarify, I think having both `partial_dependence()` and `ice()` is fine. It's the duplication of the `plot_` methods that I find potentially unnecessary)

glemaitre commented:
> +1 for a RFC (SLEP might be overkill?). We can also plan a discussion for the meeting Monday

Yes this could be a good thing.

> (Just to clarify, I think having both partial_dependence() and ice() is fine. It's the duplication of the plot_ methods that I find potentially unnecessary)

Oh ok. So basically a parameter to have `individual`, `average`, or 'both` would be potentially good. Then, my only concern is the name of the function which should indicate that you can get both PD and ICE plots out of it.

NicolasHug commented:
> Then, my only concern is the name of the function which should indicate that you can get both PD and ICE plots out of it.

Hm, yeah. Strictly peaking, ICEs *are* a form of partial dependence, so the current name could work. If we want to make it slightly more explicit, we may have the parameter `with_ice=False` by default and also allow for `with_ice=True` and `with_ice='only'`? Not super self-convinced ...

jnothman commented:
Why not just individual=False?

glemaitre commented:
> Why not just individual=False?

individual=False/True/'only'?

madhuracj committed:
Rename parameter. Add method docstring

madhuracj committed:
Update sklearn/inspection/_partial_dependence.py

Fix formatting

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

madhuracj committed:
Move _validate_parameters to the top of the file

madhuracj committed:
Merge branch 'ice' of github.com:madhuracj/scikit-learn into ice

madhuracj committed:
Use f-string formatting

madhuracj committed:
Use f-string formatting

madhuracj committed:
Jump a line

madhuracj committed:
Another f-string formatting

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -172,6 +179,95 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+        if response_method != 'decision_function':
+            raise ValueError(
+                "With the 'recursion' method, the response_method must be "
+                "'decision_function'. Got {}.".format(response_method)
```
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/b69cdef3421b57f6fbb987f8e7e61bd500a3869e

madhuracj committed:
Absorb _ice_brute method into ice method

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -401,3 +415,126 @@ def partial_dependence(estimator, X, features, response_method='auto',
     return averaged_predictions, values
 
 
+def _ice_brute(estimator, grid, features_indices, X, response_method):
```
Moved in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/c6709c15757083751bde20f0f051b4e832c09881

madhuracj committed:
Better wording, as suggested

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -401,3 +415,126 @@ def partial_dependence(estimator, X, features, response_method='auto',
+def _ice_brute(estimator, grid, features_indices, X, response_method):
+    predictions = _get_predictions(estimator, grid, features_indices, X,
+                                   response_method)
+    instances = X.shape[0]
```
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/35b19c6b2a2d683e9ba50750301a7b13bbe48d22

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -401,3 +415,126 @@ def partial_dependence(estimator, X, features, response_method='auto',
+    predictions = np.array(predictions).T
+    if is_regressor(estimator) and predictions.ndim == 2:
+        # non-multioutput regression, shape is (n_instances, n_points,)
+        predictions = predictions.reshape(instances, -1)
```
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/35b19c6b2a2d683e9ba50750301a7b13bbe48d22

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -401,3 +415,126 @@ def partial_dependence(estimator, X, features, response_method='auto',
+        # Binary classification, shape is (2, n_instances, n_points).
+        # we output the effect of **positive** class
+        predictions = predictions[1]
+        predictions = predictions.reshape(instances, -1)
```
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/35b19c6b2a2d683e9ba50750301a7b13bbe48d22

madhuracj committed:
Improve docstring

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -401,3 +415,126 @@ def partial_dependence(estimator, X, features, response_method='auto',
+        :term:`predict_proba`, or :term:`decision_function`.
+        Multioutput-multiclass classifiers are not supported.
+
+    X : {array-like or dataframe} of shape (n_samples, n_features)
```
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/980895cd40f688524e40facc465214df655d44db

madhuracj committed:
Improve formatting

madhuracj committed:
Improve docstring as suggested

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -401,3 +415,126 @@ def partial_dependence(estimator, X, features, response_method='auto',
+    # reshape predictions to
+    # (n_outputs, n_instances, n_values_feature_0, n_values_feature_1, ...)
+    predictions = predictions.reshape(
+        -1, X.shape[0], *[val.shape[0] for val in values])
```
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/c1cef8517b2de0e06728d8171cacf3e971c0f331

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -954,3 +1111,470 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
+
+    Parameters
+    ----------
+    estimator : BaseEstimator
```
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/d96de87b036ba8e39a926b1baa70150c4f6e6fd3

madhuracj committed:
Remove additional new line as suggested

madhuracj commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -172,6 +179,95 @@ def _partial_dependence_brute(est, grid, features, X, response_method):
+                .format(', '.join(supported_classes_recursion)))
+        if response_method == 'auto':
+            response_method = 'decision_function'
+
```
https://github.com/scikit-learn/scikit-learn/pull/16164/commits/4732a447ec11c2862d0fb90018fa18340c7a9af9

madhuracj committed:
Update doc/modules/partial_dependence.rst

Improve documentation explaining the reason for supporting only 'brute' for ICE

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

madhuracj committed:
pdp example should be enough given ice is similar

madhuracj commented:
Why not `plot_type = 'pdp, ice, both'` if terms 'Partial Dependence Plot' and 'Individual Conditional Expectation' is well established in the literature?

madhuracj committed:
Fix unit tests

madhuracj committed:
Merge branch 'master' into ice

glemaitre commented:
I prepared a [document](https://bit.ly/3a4tSUz) which we will discuss in the monthly meeting on Monday 24th February which was announced on the scikit-learn public mailing.

@madhuracj Feel free to join the meeting if you wish.
@NicolasHug @madhuracj @jnothman Feel free to modify the document if I forgot an important point (or if you want to shorten it :)) I think we should keep it minimal.

NicolasHug commented:
I'm not sure the discussion about efficiency is relevant? ICEs cannot be computed with the recursion method anyway since this method directly computes the averages without computing the individual values (that's why it's faster).


I'd make the discussion only about `plot_pdp(new_param='new_option')` vs `plot_pdp() + plot_ice()`. We can leave `pd()` and potential `ice()` functions out of the discussion, because they're far less useful (i.e. whatever will be fine).

- new param pros: 
  - clear that ICE is just of subcase of PD, useful from a pedagogical POV
  - easy to have both ICE and PD on the same plot (which is very useful)
- new param cons:
  - it's a new param

- 2 functions pros:
  - TODO guillaume ;)
- 2 functions cons:
  - how do we plot both PDPs and ICES on the same plot?
  - new function with 90% of code and API in common with `plot_pdp()`

Like I said before, I don't think discoverability is an issue with any of the solutions

glemaitre commented:
> I'm not sure the discussion about efficiency is relevant? ICEs cannot be computed with the recursion method anyway since this method directly computes the averages without computing the individual values (that's why it's faster).

I was probably unclear (feel free to change it), the efficiency comes from if one wants to compute ICE and PD then this would be really efficient to only average ICE while making a call to the PD function again. Basically this to avoid recomputation as your earlier pointed out.

glemaitre commented:
> I'd make the discussion only about plot_pdp(new_param='new_option') vs plot_pdp() + plot_ice(). We can leave pd() and potential ice() functions out of the discussion, because they're far less useful (i.e. whatever will be fine).

I think they come as a whole and I would prefer that everyone get the full picture in case there is a design that we did not think until now that can come up.

glemaitre commented:
> Like I said before, I don't think discoverability is an issue with any of the solutions

This is something that I personally consider an issue and that I want to raise during the discussion. This is already something similar that I defended when dealing with the `StackingEstimator`. However, if on Monday I am the only one thinking that this would be a blocker, I will be happy to drop it and go with the majority.

madhuracj committed:
Fix line length

glemaitre commented:
So following the discussion during the monthly meeting, we agree to add a parameter to the `partial_dependence` and `plot_partial_dependence` functions. I think that we can start with the parameter `individual` as proposed by @jnothman. We can then discuss when is the most sensible default for the plotting. @ogrisel was proposing to plot ICE and PD with ICE with a lower alpha and ICE on the top.

glemaitre commented:
@madhuracj Does it seem OK to go in this direction?

madhuracj commented:
Yes, I am fine with this direction. To preserve backward compatibility, the default value for `individual` should be `False` I guess while `True` draws ICEs and `'both'` draws both ICEs and PDP?

Also, be mindful that the output shape of `partial_dependence` will be dependant of this parameter.
- `False`: averaged_predictions : ndarray, shape (n_outputs, len(values[0]), len(values[1]), ...)
- `True`: predictions : ndarray, shape (n_outputs, n_instances, len(values[0]), len(values[1]),...) 
- `'both'`: (averaged_predictions, predictions): tuple

I should be able to work on the changes during the next weekend.

JoElfner commented:
Any news on this, especially the `centre` keyword? If this PR is delayed or deprecated, I could commit a quick fix for centered ICE plots.

glemaitre commented:
It has been implemented in #16619 as stated just on the line above your comment. You can check the documentation:
https://scikit-learn.org/dev/modules/generated/sklearn.inspection.plot_partial_dependence.html#sklearn.inspection.plot_partial_dependence

It can be used with the parameter `kind` and you can check the examples for illustration.

JoElfner commented:
@glemaitre 
Thank you for your answer.
I know. I've seen this (and other) PR, and have checked all changed files (also PRs which are not yet merged). But as I wrote, I am specifically looking for the `center`/`centre` argument/keyword, producing centered ICE-plots (also called cICE or c-ICE plots).
Good examples can be found [here in section 5.2.1.1](https://christophm.github.io/interpretable-ml-book/ice.html) or [here](http://savvastjortjoglou.com/intrepretable-machine-learning-nfl-combine.html#Centered-ICE-Plots).
Since the implementation is fairly easy already having non-centered ICE-plots, I could commit a quick PR adding the possibility to center the ICE plots, as in commit 2f85ed3 of this PR (which seems to be deprecated).

glemaitre commented:
Open an issue first such that we discuss the API and if we want to support this feature or not.

JoElfner commented:
Centering ICE plots is, in my opinion and possibly only speaking for the field of ML applied to engineering topics, the parameter increasing the explanatory power of ICE plots the most.
I'd even go so far as to say: Not supporting centered ICE plots is like building a car but not supporting wheels.
Meaning you already implemented the complex and important trunk, but you would miss out on the easy-to-implement but powerful feature.
But ok, I'll open an issue so we can discuss the API in detail.
